Zed does not respect mouse cursor size

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I have the mouse cursor on my desktop set to medium size (Settings/Accessibility/Seeing/Cursor Size) but it's the default size in Zed.

### Environment

Zed: v0.146.5 (Zed)
OS: Linux Wayland fedora 40
Memory: 31.2 GiB
Architecture: x86_64
GPU: AMD Radeon RX 7800 XT (RADV NAVI32) || radv || Mesa 24.1.5

### If applicable, add mockups / screenshots to help explain present your vision of the feature

Here's how the cursor looks on my desktop.
![cursor_de](https://github.com/user-attachments/assets/c0d09dac-d2d0-4275-8b3d-e61a6a848344)

Here's how the cursor looks in Zed.
![cursor_zed](https://github.com/user-attachments/assets/5af4f99e-a35a-4281-be50-e7c4b63dea28)


### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>
<!-- Click below this line and paste or drag-and-drop your log-->

<!-- Click above this line and paste or drag-and-drop your log--></pre></details>

> * `gdbus call -e -d org.freedesktop.portal.Desktop -o /org/freedesktop/portal/desktop -m org.freedesktop.portal.Settings.Read '"org.gnome.desktop.interface"' '"cursor-theme"'`

(<<'Mocu-Black-Right'>>,)

> * `gdbus call -e -d org.freedesktop.portal.Desktop -o /org/freedesktop/portal/desktop -m org.freedesktop.portal.Settings.Read '"org.gnome.desktop.interface"' '"cursor-size"'`

(<<32>>,)

> And could you share your logs? They may contain some important debugging information.

<details><summary>Zed.log</summary><pre>
<!-- Click below this line and paste or drag-and-drop your log-->
2024-08-05T17:21:36.685430837-07:00 [INFO] ========== starting zed ==========
2024-08-05T17:21:36.707942895-07:00 [INFO] perform;
2024-08-05T17:21:36.708157366-07:00 [INFO] Opening main db
2024-08-05T17:21:36.708286348-07:00 [INFO] read_command;
2024-08-05T17:21:36.708520035-07:00 [INFO] read_command;
2024-08-05T17:21:36.708910436-07:00 [INFO] socket reader;
2024-08-05T17:21:36.70935607-07:00 [INFO] new;
2024-08-05T17:21:36.710032898-07:00 [INFO] keep_updated;
2024-08-05T17:21:36.710510121-07:00 [INFO] Using git binary path: None
2024-08-05T17:21:36.741035132-07:00 [INFO] extensions updated. loading 7, reloading 0, unloading 0
2024-08-05T17:21:36.743006764-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:36.743137318-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:36.743247514-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:36.743369783-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:36.743898343-07:00 [INFO] activate is not implemented on Linux, ignoring the call
2024-08-05T17:21:36.74404619-07:00 [INFO] Opening main db
2024-08-05T17:21:36.745528425-07:00 [INFO] perform;
2024-08-05T17:21:36.745699776-07:00 [INFO] read_command;
2024-08-05T17:21:36.745825812-07:00 [INFO] read_command;
2024-08-05T17:21:36.745986222-07:00 [INFO] socket reader;
2024-08-05T17:21:36.750824931-07:00 [INFO] set environment variables from shell:/bin/bash, path:/home/barugon/.cargo/bin:/home/barugon/.local/bin:/home/barugon/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin
2024-08-05T17:21:36.759083414-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:36.759205633-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:36.759313224-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:36.759420725-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:36.759549557-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:36.783640251-07:00 [INFO] Enabling Vulkan Portability
2024-08-05T17:21:36.783763922-07:00 [INFO] Enabling color space support
2024-08-05T17:21:36.800877614-07:00 [INFO] Adapter "AMD Radeon RX 7800 XT (RADV NAVI32)"
2024-08-05T17:21:36.801023858-07:00 [INFO] Ray tracing is supported
2024-08-05T17:21:37.103618207-07:00 [INFO] Using surface present mode MAILBOX
2024-08-05T17:21:37.104206067-07:00 [WARN] Unable to forbid exclusive full screen
2024-08-05T17:21:37.105270059-07:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-08-05T17:21:37.105405843-07:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-08-05T17:21:37.10554272-07:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: PreMultiplied }
2024-08-05T17:21:37.111414343-07:00 [INFO] Using GPU: GPUSpecs { is_software_emulated: false, device_name: "AMD Radeon RX 7800 XT (RADV NAVI32)", driver_name: "radv", driver_info: "Mesa 24.1.5" }
2024-08-05T17:21:37.112828681-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:37.113012776-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:37.114221389-07:00 [INFO] Opening main db
2024-08-05T17:21:37.196370644-07:00 [INFO] Initializing default prettier with plugins {}
2024-08-05T17:21:37.196950109-07:00 [ERROR] no worktrees when constructing ProjectLspAdapterDelegate
2024-08-05T17:21:37.203642419-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:37.224483149-07:00 [INFO] Using surface present mode MAILBOX
2024-08-05T17:21:37.225289678-07:00 [WARN] Unable to forbid exclusive full screen
2024-08-05T17:21:37.231709298-07:00 [INFO] Initializing default prettier with plugins {}
2024-08-05T17:21:37.236673552-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:37.499232464-07:00 [INFO] Initializing default prettier with plugins {}
2024-08-05T17:21:37.500729517-07:00 [WARN] request completed with error: failed to connect to the server
2024-08-05T17:21:40.568874926-07:00 [INFO] Using surface present mode MAILBOX
2024-08-05T17:21:40.56963535-07:00 [WARN] Unable to forbid exclusive full screen
2024-08-05T17:21:42.419094134-07:00 [INFO] Using surface present mode MAILBOX
2024-08-05T17:21:42.419821787-07:00 [WARN] Unable to forbid exclusive full screen
<!-- Click above this line and paste or drag-and-drop your log--></pre></details>

--- a/crates/gpui/src/platform/linux/wayland/client.rs
+++ b/crates/gpui/src/platform/linux/wayland/client.rs
@@ -496,7 +496,7 @@ impl WaylandClient {
                     XDPEvent::CursorTheme(theme) => {
                         if let Some(client) = client.0.upgrade() {
                             let mut client = client.borrow_mut();
-                            client.cursor.set_theme(theme.as_str(), None);
+                            client.cursor.set_theme(theme.as_str());
                         }
                     }
                     XDPEvent::CursorSize(size) => {
@@ -649,15 +649,16 @@ impl LinuxClient for WaylandClient {
 
             if let Some(cursor_shape_device) = &state.cursor_shape_device {
                 cursor_shape_device.set_shape(serial, style.to_shape());
-            } else if state.mouse_focused_window.is_some() {
+            } else if let Some(focused_window) = &state.mouse_focused_window {
                 // cursor-shape-v1 isn't supported, set the cursor using a surface.
                 let wl_pointer = state
                     .wl_pointer
                     .clone()
                     .expect("window is focused by pointer");
+                let scale = focused_window.primary_output_scale();
                 state
                     .cursor
-                    .set_icon(&wl_pointer, serial, &style.to_icon_name());
+                    .set_icon(&wl_pointer, serial, &style.to_icon_name(), scale);
             }
         }
     }
@@ -1439,9 +1440,13 @@ impl Dispatch<wl_pointer::WlPointer, ()> for WaylandClientStatePtr {
                         if let Some(cursor_shape_device) = &state.cursor_shape_device {
                             cursor_shape_device.set_shape(serial, style.to_shape());
                         } else {
-                            state
-                                .cursor
-                                .set_icon(&wl_pointer, serial, &style.to_icon_name());
+                            let scale = window.primary_output_scale();
+                            state.cursor.set_icon(
+                                &wl_pointer,
+                                serial,
+                                &style.to_icon_name(),
+                                scale,
+                            );
                         }
                     }
                     drop(state);

--- a/crates/gpui/src/platform/linux/wayland/cursor.rs
+++ b/crates/gpui/src/platform/linux/wayland/cursor.rs
@@ -9,6 +9,7 @@ use wayland_cursor::{CursorImageBuffer, CursorTheme};
 pub(crate) struct Cursor {
     theme: Option<CursorTheme>,
     theme_name: Option<String>,
+    theme_size: u32,
     surface: WlSurface,
     size: u32,
     shm: WlShm,
@@ -27,6 +28,7 @@ impl Cursor {
         Self {
             theme: CursorTheme::load(&connection, globals.shm.clone(), size).log_err(),
             theme_name: None,
+            theme_size: size,
             surface: globals.compositor.create_surface(&globals.qh, ()),
             shm: globals.shm.clone(),
             connection: connection.clone(),
@@ -34,26 +36,26 @@ impl Cursor {
         }
     }
 
-    pub fn set_theme(&mut self, theme_name: &str, size: Option<u32>) {
-        if let Some(size) = size {
-            self.size = size;
-        }
-        if let Some(theme) =
-            CursorTheme::load_from_name(&self.connection, self.shm.clone(), theme_name, self.size)
-                .log_err()
+    pub fn set_theme(&mut self, theme_name: &str) {
+        if let Some(theme) = CursorTheme::load_from_name(
+            &self.connection,
+            self.shm.clone(),
+            theme_name,
+            self.theme_size,
+        )
+        .log_err()
         {
             self.theme = Some(theme);
             self.theme_name = Some(theme_name.to_string());
         } else if let Some(theme) =
-            CursorTheme::load(&self.connection, self.shm.clone(), self.size).log_err()
+            CursorTheme::load(&self.connection, self.shm.clone(), self.theme_size).log_err()
         {
             self.theme = Some(theme);
             self.theme_name = None;
         }
     }
 
-    pub fn set_size(&mut self, size: u32) {
-        self.size = size;
+    fn set_theme_size(&mut self, theme_size: u32) {
         self.theme = self
             .theme_name
             .as_ref()
@@ -62,14 +64,29 @@ impl Cursor {
                     &self.connection,
                     self.shm.clone(),
                     name.as_str(),
-                    self.size,
+                    theme_size,
                 )
                 .log_err()
             })
-            .or_else(|| CursorTheme::load(&self.connection, self.shm.clone(), self.size).log_err());
+            .or_else(|| {
+                CursorTheme::load(&self.connection, self.shm.clone(), theme_size).log_err()
+            });
     }
 
-    pub fn set_icon(&mut self, wl_pointer: &WlPointer, serial_id: u32, mut cursor_icon_name: &str) {
+    pub fn set_size(&mut self, size: u32) {
+        self.size = size;
+        self.set_theme_size(size);
+    }
+
+    pub fn set_icon(
+        &mut self,
+        wl_pointer: &WlPointer,
+        serial_id: u32,
+        mut cursor_icon_name: &str,
+        scale: i32,
+    ) {
+        self.set_theme_size(self.size * scale as u32);
+
         if let Some(theme) = &mut self.theme {
             let mut buffer: Option<&CursorImageBuffer>;
 
@@ -91,7 +108,15 @@ impl Cursor {
                 let (width, height) = buffer.dimensions();
                 let (hot_x, hot_y) = buffer.hotspot();
 
-                wl_pointer.set_cursor(serial_id, Some(&self.surface), hot_x as i32, hot_y as i32);
+                self.surface.set_buffer_scale(scale);
+
+                wl_pointer.set_cursor(
+                    serial_id,
+                    Some(&self.surface),
+                    hot_x as i32 / scale,
+                    hot_y as i32 / scale,
+                );
+
                 self.surface.attach(Some(&buffer), 0, 0);
                 self.surface.damage(0, 0, width as i32, height as i32);
                 self.surface.commit();

--- a/crates/gpui/src/platform/linux/wayland/window.rs
+++ b/crates/gpui/src/platform/linux/wayland/window.rs
@@ -194,6 +194,23 @@ impl WaylandWindowState {
         self.decorations == WindowDecorations::Client
             || self.background_appearance != WindowBackgroundAppearance::Opaque
     }
+
+    pub fn primary_output_scale(&mut self) -> i32 {
+        let mut scale = 1;
+        let mut current_output = self.display.take();
+        for (id, output) in self.outputs.iter() {
+            if let Some((_, output_data)) = &current_output {
+                if output.scale > output_data.scale {
+                    current_output = Some((id.clone(), output.clone()));
+                }
+            } else {
+                current_output = Some((id.clone(), output.clone()));
+            }
+            scale = scale.max(output.scale);
+        }
+        self.display = current_output;
+        scale
+    }
 }
 
 pub(crate) struct WaylandWindow(pub WaylandWindowStatePtr);
@@ -560,7 +577,7 @@ impl WaylandWindowStatePtr {
 
                 state.outputs.insert(id, output.clone());
 
-                let scale = primary_output_scale(&mut state);
+                let scale = state.primary_output_scale();
 
                 // We use `PreferredBufferScale` instead to set the scale if it's available
                 if state.surface.version() < wl_surface::EVT_PREFERRED_BUFFER_SCALE_SINCE {
@@ -572,7 +589,7 @@ impl WaylandWindowStatePtr {
             wl_surface::Event::Leave { output } => {
                 state.outputs.remove(&output.id());
 
-                let scale = primary_output_scale(&mut state);
+                let scale = state.primary_output_scale();
 
                 // We use `PreferredBufferScale` instead to set the scale if it's available
                 if state.surface.version() < wl_surface::EVT_PREFERRED_BUFFER_SCALE_SINCE {
@@ -719,6 +736,10 @@ impl WaylandWindowStatePtr {
             (fun)()
         }
     }
+
+    pub fn primary_output_scale(&self) -> i32 {
+        self.state.borrow_mut().primary_output_scale()
+    }
 }
 
 fn extract_states<'a, S: TryFrom<u32> + 'a>(states: &'a [u8]) -> impl Iterator<Item = S> + 'a
@@ -730,23 +751,6 @@ where
         .flat_map(TryInto::<[u8; 4]>::try_into)
         .map(u32::from_ne_bytes)
         .flat_map(S::try_from)
-}
-
-fn primary_output_scale(state: &mut RefMut<WaylandWindowState>) -> i32 {
-    let mut scale = 1;
-    let mut current_output = state.display.take();
-    for (id, output) in state.outputs.iter() {
-        if let Some((_, output_data)) = &current_output {
-            if output.scale > output_data.scale {
-                current_output = Some((id.clone(), output.clone()));
-            }
-        } else {
-            current_output = Some((id.clone(), output.clone()));
-        }
-        scale = scale.max(output.scale);
-    }
-    state.display = current_output;
-    scale
 }
 
 impl rwh::HasWindowHandle for WaylandWindow {

--- a/crates/gpui/src/platform/linux/xdg_desktop_portal.rs
+++ b/crates/gpui/src/platform/linux/xdg_desktop_portal.rs
@@ -42,11 +42,13 @@ impl XDPEventSource {
                 {
                     sender.send(Event::CursorTheme(initial_theme))?;
                 }
+
+                // If u32 is used here, it throws invalid type error
                 if let Ok(initial_size) = settings
-                    .read::<u32>("org.gnome.desktop.interface", "cursor-size")
+                    .read::<i32>("org.gnome.desktop.interface", "cursor-size")
                     .await
                 {
-                    sender.send(Event::CursorSize(initial_size))?;
+                    sender.send(Event::CursorSize(initial_size as u32))?;
                 }
 
                 if let Ok(mut cursor_theme_changed) = settings
@@ -69,7 +71,7 @@ impl XDPEventSource {
                 }
 
                 if let Ok(mut cursor_size_changed) = settings
-                    .receive_setting_changed_with_args::<u32>(
+                    .receive_setting_changed_with_args::<i32>(
                         "org.gnome.desktop.interface",
                         "cursor-size",
                     )
@@ -80,7 +82,7 @@ impl XDPEventSource {
                         .spawn(async move {
                             while let Some(size) = cursor_size_changed.next().await {
                                 let size = size?;
-                                sender.send(Event::CursorSize(size))?;
+                                sender.send(Event::CursorSize(size as u32))?;
                             }
                             anyhow::Ok(())
                         })

